iroh/src/commands/sync.rs

```diff
@@ -354,3 +480,273 @@ async fn print_entry(doc: &Doc, entry: &Entry, content: bool) -> anyhow::Result<
+                                Err(e) => {
+                                    doc_set_updates
+                                        .send(DocSetStreamUpdate::Abort)
+                                        .await
+                                        .expect("receiver dropped");
+                                    anyhow::bail!("issue creating key for entry {hash:?}: {e}");
+                                }
+                            };
+                            // send update to doc
+                            doc_set_updates
+                                .send(DocSetStreamUpdate::Entry {
+                                    id,
+                                    key,
+                                    hash,
+                                    size: *size,
+                                })
+                                .await
+                                .expect("receiver dropped");
+                        }
+                        None => {
+                            anyhow::bail!("Got Done for unknown collection id {id}");
+                        }
+                    }
+                }
+                AddProgress::AllDone { hash, .. } => {
+                    tracing::trace!("AddProgress::AllDone({hash:?})");
+                    break;
+                }
+                AddProgress::Abort(e) => {
+                    task_mp.error().await;
+                    doc_set_updates
+                        .send(DocSetStreamUpdate::Abort)
+                        .await
+                        .expect("receiver dropped");
+                    anyhow::bail!("Error while adding data: {e}");
+                }
+            }
+        }
+        Ok(())
+    });
+    let mut entries = 0;
+    let mut total_size = 0;
+    while let Some(res) = doc_set_progress.next().await {
+        match res? {
+            DocSetProgress::Done { id, size, .. } => {
+                mp.done(id).await;
+                entries += 1;
+                total_size += size;
+            }
+            DocSetProgress::AllDone => {
+                mp.all_done().await;
+                add_progress_task.abort();
+                return Ok((entries, total_size));
+            }
+            DocSetProgress::Abort(e) => {
+                mp.error().await;
+                add_progress_task.abort();
+                anyhow::bail!("Error while adding entry to doc: {e}")
+            }
+        }
+    }
+    unreachable!();
+}
+
+fn key_from_path_str(root: PathBuf, prefix: String, path_str: String) -> Result<Vec<u8>> {
+    let suffix = PathBuf::from(path_str)
+        .strip_prefix(root)?
+        .to_str()
+        .map(|p| p.as_bytes())
+        .ok_or(anyhow!("could not convert path to bytes"))?
+        .to_vec();
+    let mut key = prefix.into_bytes().to_vec();
+    key.extend(suffix);
+    Ok(key)
+}
+
+#[derive(Debug, Clone)]
+struct ImportProgressState {
+    task: Arc<Mutex<Option<JoinHandle<()>>>>,
```

Comment:
Can you try taking out the join handle here, so that ImportProgressState is basically just a newtype for the sender?

new would then return Self and an AbortingJoinHandle<()>, and you just have to keep the latter around until you're done. That would allow getting rid of both the Arc and the Mutex.

Reply:
that's so much better, thank you for the suggestion!!!